lib/reporting/protocols_report.rb

```diff
@@ -0,0 +1,196 @@
+        Reporting::EmailableReport.new(
+          title: 'SAML Signature Issues',
+          table: saml_signature_issues_table,
+        ),
+      ]
+    end
+
+    def to_csvs
+      as_tables.map do |table|
+        CSV.generate do |csv|
+          table.each do |row|
+            csv << row
+          end
+        end
+      end
+    end
+
+    # @return Array<Hash>
+    def data
+      @data ||= begin
+        fetch_results
+      end
+    end
+
+    def fetch_results
+      cloudwatch_client.fetch(query:, from: time_range.begin, to: time_range.end)
+    end
+
+    def query
+      params = {
+        event: quote([SAML_AUTH_EVENT, OIDC_AUTH_EVENT]),
+      }
+
+      format(<<~QUERY, params)
+        fields
+          name as protocol,
+          properties.event_properties.service_provider as issuer,
+          properties.event_properties.request_signed = 1 AS signed,
+          properties.event_properties.request_signed != 1 AS not_signed,
+          isempty(properties.event_properties.matching_cert_serial) and signed as invalid_signature
+        | filter name IN %{event}
+        | stats
+          count(*) as request_count,
+          sum(signed) as signed_count,
+          sum(not_signed) as unsigned_count,
+          sum(invalid_signature) as invalid_signature_count
+          by
+          protocol,
+          issuer
+      QUERY
+    end
+
+    def cloudwatch_client
+      @cloudwatch_client ||= Reporting::CloudwatchClient.new(
+        num_threads: @threads,
+        ensure_complete_logs: true,
+        slice_interval: @slice,
+        progress: progress?,
+        logger: verbose? ? Logger.new(STDERR) : nil,
+      )
+    end
+
+    def overview_table
+      [
+        ['Report Timeframe', "#{time_range.begin} to #{time_range.end}"],
+        # This needs to be Date.today so it works when run on the command line
+        ['Report Generated', Date.today.to_s], # rubocop:disable Rails/Date
+      ]
+    end
+
+    def sum_by(key_name, key_value, metric_name)
+      data.inject(0) do |sum, slice|
+        sum + (slice[key_name] == key_value ? slice[metric_name].to_i : 0)
+      end
+    end
+
+    def protocols_table
+      saml_total = sum_by('protocol', SAML_AUTH_EVENT, 'request_count')
+      oidc_total = sum_by('protocol', OIDC_AUTH_EVENT, 'request_count')
+
+      [
+        ['Authentication Protocol', '% of attempts', 'Total number'],
+        [
+          'SAML',
+          to_percent(saml_total, saml_total + oidc_total),
+          saml_total,
+        ],
+        [
+          'OIDC',
+          to_percent(oidc_total, saml_total + oidc_total),
+          oidc_total,
+        ],
+      ]
+    end
+
+    def saml_signature_issues_table
+      issuers_with_unsigned_requests = data.collect do |slice|
+        slice['unsigned_count'].to_i > 0 ? slice['issuer'] : nil
+      end.compact.uniq
+      issuers_with_invalid_signatures = data.collect do |slice|
+        slice['invalid_signature_count'].to_i > 0 ? slice['issuer'] : nil
```

Comment:
should `nil` be `[]`? (also on line 162)

Reply:
It is `nil` so that `Array#compact` can remove it. I don't know if this is idiomatic Ruby. I miss my list comprehensions.

Reply:
oh i see, i misread what was happening here. i think it would be easier/more readable to use [keep_if](https://apidock.com/ruby/Array/keep_if) than `collect`.

```ruby
issuers_with_invalid_signatures = data.keep_if { |slice| slice['invalid_signature_count'].to_i > 0 }.uniq
```

Reply:
That is still not enough. I would still need to do `collect` or `map`. It would have to be:
```
issuers_with_invalid_signatures = data
  .keep_if { |slice| slice['invalid_signature_count'].to_i > 0 }
  .collect { |slice| slice['issuer'] }
  .uniq
```

Would you prefer that?

Reply:
personally I like `select` and `map` but otherwise same same
```ruby
issuers_with_invalid_signatures = data
  .select { |slice| slice['invalid_signature_count'].to_i > 0 }
  .map { |slice| slice['issuer'] }
  .uniq
```

Reply:
Strange, Rubocop wants "Place the . on the previous line, together with the method call receiver."

Reply:
oh yeah, i don't like that particular rubocop option, but it's what we have standardized